nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/StandardAnalysisContext.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.atlas.provenance;
+
+import org.apache.nifi.atlas.NiFiFlow;
+import org.apache.nifi.atlas.resolver.ClusterResolver;
+import org.apache.nifi.authorization.user.NiFiUser;
+import org.apache.nifi.controller.status.ConnectionStatus;
+import org.apache.nifi.provenance.ProvenanceEventRecord;
+import org.apache.nifi.provenance.ProvenanceRepository;
+import org.apache.nifi.provenance.lineage.ComputeLineageResult;
+import org.apache.nifi.provenance.lineage.ComputeLineageSubmission;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+public class StandardAnalysisContext implements AnalysisContext {
+
+    private final Logger logger = LoggerFactory.getLogger(StandardAnalysisContext.class);
+    private final NiFiFlow nifiFlow;
+    private final ClusterResolver clusterResolver;
+    private final ProvenanceRepository provenanceRepository;
+
+    public StandardAnalysisContext(NiFiFlow nifiFlow, ClusterResolver clusterResolver,
+                                   ProvenanceRepository provenanceRepository) {
+        this.nifiFlow = nifiFlow;
+        this.clusterResolver = clusterResolver;
+        this.provenanceRepository = provenanceRepository;
+    }
+
+    @Override
+    public List<ConnectionStatus> findConnectionTo(String componentId) {
+        return nifiFlow.getIncomingRelationShips(componentId);
+    }
+
+    @Override
+    public List<ConnectionStatus> findConnectionFrom(String componentId) {
+        return nifiFlow.getOutgoingRelationShips(componentId);
+    }
+
+    @Override
+    public String getNiFiClusterName() {
+        return nifiFlow.getClusterName();
+    }
+
+    @Override
+    public ClusterResolver getClusterResolver() {
+        return clusterResolver;
+    }
+
+    private ComputeLineageResult getLineageResult(long eventId, ComputeLineageSubmission submission) {
+        final ComputeLineageResult result = submission.getResult();
+        try {
+            if (result.awaitCompletion(10, TimeUnit.SECONDS)) {
```

Comment:
I'm not sure yet the context in which this is used. However, this is a bit of a red flag. In order to compute a FlowFile's lineage, depending on the size of the Provenance Repository and the implementation, it may take many minutes (consider a Provenance Repository that is a couple of terabytes in size and the event is one of the oldest -it will have to search all lucene indices for this). In most cases it is fairly quick, but we should not assume that it will be. What exactly is the implication here if this always times out?

Reply:
I agree on that this can be a costly operation. The reason to query provenance is to compute a lineage from a DROP provenance event. This is used by 'Complete Path' strategy. I wrote documentation on performance impact. If this does not work for a use-case, then user can choose another strategy, 'Simple Path'. Simple Path does not query provenance events this way. It analyzes each individual event, so should be more lightweight.